src/v/utils/tracking_allocator.h

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright 2020 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "utils/intrusive_list_helpers.h"
+
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/sstring.hh>
+
+#include <memory>
+#include <vassert.h>
+
+namespace ss = seastar;
+
+namespace util {
+
+/// Tracks allocations and deallocations of memory. A mem_tracker can have
+/// child trackers effectively forming an n-ary tree like structure.
+class mem_tracker {
+public:
+    explicit mem_tracker(ss::sstring label)
+      : _label(std::move(label)) {}
+    mem_tracker(mem_tracker&) = delete;
+    mem_tracker(mem_tracker&&) = default;
+    mem_tracker& operator=(mem_tracker&) = delete;
+    mem_tracker& operator=(mem_tracker&&) = delete;
+    ~mem_tracker() = default;
+
+    void allocate(int64_t bytes) { _consumption += bytes; }
+
+    void deallocate(int64_t bytes) { _consumption -= bytes; }
+
+    ss::shared_ptr<mem_tracker> create_child(ss::sstring);
```

Comment:
It may be worth changing the interface(or just adding a new method) here to something like `void register_child(mem_tracker& child);` this would allow the caller to allocate a child into whatever smart pointer type they want without having to move it out of the returned `shared_ptr`. I can see some users not needing the reference counting.

Reply:
I think having a shared_ptr for children makes the ownership of the child tracker clear, that it is shared. Also a shared_ptr to make it thread-safe (see other comment comment from Travis) and make the allocator itself copyable and movable (or may be I'm missing something, happy to discuss).